core/src/main/scala/com/pingcap/tispark/v2/TiDBTable.scala

```diff
@@ -0,0 +1,183 @@
+import org.apache.spark.sql.connector.catalog.{SupportsRead, TableCapability}
+import org.apache.spark.sql.connector.read.ScanBuilder
+import org.apache.spark.sql.execution.{ColumnarCoprocessorRDD, SparkPlan}
+import org.apache.spark.sql.tispark.{TiHandleRDD, TiRowRDD}
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+import org.apache.spark.sql.{SQLContext, execution}
+
+import java.util
+import java.util.Collections
+import scala.collection.mutable.ListBuffer
+import collection.JavaConverters._
+
+case class TiDBTable(
+    session: TiSession,
+    tableRef: TiTableReference,
+    meta: MetaManager,
+    var ts: TiTimestamp = null,
+    options: Option[TiDBOptions] = None)(@transient val sqlContext: SQLContext)
+    extends SupportsRead {
+
+  implicit class IdentifierHelper(identifier: TiTableReference) {
+    def quoted: String = {
+      Seq(identifier.databaseName, identifier.tableName).map(quote).mkString(".")
+    }
+    private def quote(part: String): String = {
+      if (part.contains(".") || part.contains("`")) {
+        s"`${part.replace("`", "``")}`"
+      } else {
+        part
+      }
+    }
+  }
+
+  lazy val table: TiTableInfo = meta
+    .getTable(tableRef.databaseName, tableRef.tableName)
+    .getOrElse(throw new NoSuchTableException(tableRef.databaseName, tableRef.tableName))
+
+  override lazy val schema: StructType = TiUtil.getSchemaFromTable(table)
+
+  override lazy val properties: util.Map[String, String] = {
+    if (options.isEmpty) {
+      Collections.emptyMap()
+    } else {
+      options.get.parameters.toMap.asJava
+    }
+  }
+
+  lazy val isTiFlashReplicaAvailable: Boolean = {
+    // Note:
+    // - INFORMATION_SCHEMA.TIFLASH_REPLICA is not present in TiKV or PD,
+    // it is calculated in TiDB and stored in memory.
+    // - In order to get those helpful information we have to read them from
+    // either TiKV or PD and keep them in memory as well.
+    //
+    // select * from INFORMATION_SCHEMA.TIFLASH_REPLICA where table_id = $id
+    // TABLE_SCHEMA, TABLE_NAME, TABLE_ID, REPLICA_COUNT, LOCATION_LABELS, AVAILABLE, PROGRESS
+    table.getTiflashReplicaInfo != null && table.getTiflashReplicaInfo.isAvailable
+  }
+
+  def databaseName: String = tableRef.databaseName
+  def tableName: String = tableRef.tableName
+
+  override def newScanBuilder(options: CaseInsensitiveStringMap): ScanBuilder =
+    () => () => schema
+
+  override def name(): String = tableRef.quoted
+
+  override def capabilities(): util.Set[TableCapability] = {
+    val capabilities = new util.HashSet[TableCapability]
+    capabilities.add(TableCapability.BATCH_READ)
+    capabilities
+  }
+
+  override def toString: String = s"TiDBTable($name)"
+
+  def dagRequestToRegionTaskExec(dagRequest: TiDAGRequest, output: Seq[Attribute]): SparkPlan = {
+    getDagRequestToRegionTaskExec(dagRequest, output, session, sqlContext, tableRef)
```

Comment:
why don't move the session and these two method to object

Reply:
Maybe session can be moved out. but this two method can't because they also need sqlContext and tableRef